Tests/Simulation/SimulatedLoadTests.cs

```diff
@@ -123,6 +127,99 @@ async Task ActAsync(SimulatedDevice device, TimeSpan startOffset)
             }
         }
 
+        /// <summary>
+        /// This test emulates a scenario where we have several plants with multiple concentrators and devices,
+        /// connected to several gateways (either on-premises or in the cloud).
+        /// </summary>
+        [Fact]
+        public async Task Connected_Factory_Load_Test_Scenario()
+        {
+            const int numberOfFactories = 1;
+            const double joinsPerSecond = 1.5;
+            var messagesPerSecond = MoreLinq.MoreEnumerable.Generate(1.5, old => old > 10 ? old : old + 2);
+            const int numberOfLoops = 5;
+            var stationsPerFactory = this.simulatedBasicsStations.Count / numberOfFactories;
+
+            // The total number of concentratos can be configured via the test configuration. It will de distributed evenly among factories;
+            // The total number of devices can be configured via the test configuration. It will de distributed evenly among factories.
+
+            // Devices are distributed across different LNS within the same factory to "increase resiliency". In the case of two LNS and one factory,
+            // this means that for two concentrators, concentrator 1 will be connected to LNS 1 while concentrator 2 will talk to LNS 2.
+            // This influences how many messages we expect in IoT Hub (due to different concentrator/LNS deduplication strategies).
+
+            Assert.True(this.simulatedBasicsStations.Count >= numberOfFactories, "There needs to be at least one concentrator per factory.");
+            Assert.True(stationsPerFactory % Configuration.LnsEndpointsForSimulator.Count == 0, "LNS must be distributed evenly across factories (identical amount of indirectly connected LNS to factories).");
+
+            var testDeviceInfo = TestFixtureSim.DeviceRange9000_OTAA_FullLoad_DuplicationDrop;
+            var devicesAndConcentratorsByFactory =
+                this.simulatedBasicsStations.Batch(stationsPerFactory)
```

Comment:
.NET 6 added [`Chunk`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.chunk?view=net-6.0), which is the same as [`Batch` from MoreLINQ](https://morelinq.github.io/3.3/ref/api/html/M_MoreLinq_MoreEnumerable_Batch__1.htm), so the latter can be simply replaced with the former:

```diff
diff --git a/Tests/Simulation/SimulatedLoadTests.cs b/Tests/Simulation/SimulatedLoadTests.cs
index f904f6eb..b2d343c3 100644
--- a/Tests/Simulation/SimulatedLoadTests.cs
+++ b/Tests/Simulation/SimulatedLoadTests.cs
@@ -17,7 +17,7 @@ namespace LoRaWan.Tests.Simulation
     using Microsoft.Extensions.Logging;
     using Xunit;
     using Xunit.Abstractions;
-    using static MoreLinq.Extensions.BatchExtension;
+    using static MoreLinq.Extensions.RepeatExtension;
     using static MoreLinq.Extensions.IndexExtension;
     using static MoreLinq.Extensions.TransposeExtension;
 
@@ -42,8 +42,9 @@ namespace LoRaWan.Tests.Simulation
             this.uniqueMessageFragment = Guid.NewGuid().ToString();
             this.logger = new TestOutputLogger(testOutputHelper);
             this.simulatedBasicsStations =
-                testFixture.DeviceRange5000_BasicsStationSimulators.Index()
-                           .Select(b => new SimulatedBasicsStation(StationEui.Parse(b.Value.DeviceID), Configuration.LnsEndpointsForSimulator[b.Key % Configuration.LnsEndpointsForSimulator.Count]))
+                testFixture.DeviceRange5000_BasicsStationSimulators
+                           .Zip(Configuration.LnsEndpointsForSimulator.Repeat(),
+                                (tdi, lnsUrl) => new SimulatedBasicsStation(StationEui.Parse(tdi.DeviceID), lnsUrl))
                            .ToList();
 
             Assert.True(this.simulatedBasicsStations.Count % Configuration.LnsEndpointsForSimulator.Count == 0, "Since Basics Stations are round-robin distributed to LNS, we must have the same number of stations per LNS for well-defined test assertions.");
@@ -152,9 +153,9 @@ namespace LoRaWan.Tests.Simulation
 
             var testDeviceInfo = TestFixtureSim.DeviceRange9000_OTAA_FullLoad_DuplicationDrop;
             var devicesAndConcentratorsByFactory =
-                this.simulatedBasicsStations.Batch(stationsPerFactory)
+                this.simulatedBasicsStations.Chunk(stationsPerFactory)
                                             .Take(numberOfFactories)
-                                            .Zip(testDeviceInfo.Batch(testDeviceInfo.Count / numberOfFactories).Take(numberOfFactories))
+                                            .Zip(testDeviceInfo.Chunk(testDeviceInfo.Count / numberOfFactories).Take(numberOfFactories))
                                             .Select(b => (Stations: b.First.ToList(), DeviceInfo: b.Second))
                                             .Select(b => (b.Stations, Devices: b.DeviceInfo.Select(d => InitializeSimulatedDevice(d, b.Stations)).ToList()))
                                             .Index()
@@ -242,7 +243,7 @@ namespace LoRaWan.Tests.Simulation
             Assert.True(messagesBeforeConfirmed <= messagesBeforeJoin, "OTAA devices should send all messages as confirmed messages.");
 
             // 1. picking devices send an initial message (warm device cache in LNS module)
-            foreach (var devices in simulatedAbpDevices.Batch(batchSizeWarmupMessages))
+            foreach (var devices in simulatedAbpDevices.Chunk(batchSizeWarmupMessages))
             {
                 await Task.WhenAll(from device in devices select SendWarmupMessageAsync(device));
                 await Task.Delay(warmupDelay);
@@ -262,7 +263,7 @@ namespace LoRaWan.Tests.Simulation
             for (var messageId = initialMessageId; messageId < initialMessageId + messagesPerDeviceExcludingWarmup; ++messageId)
             {
                 foreach (var batch in simulatedAbpDevices.Zip(simulatedOtaaDevices, (first, second) => (Abp: first, Otaa: second))
-                                                         .Batch(batchSizeDataMessages))
+                                                         .Chunk(batchSizeDataMessages))
                 {
                     var payload = this.uniqueMessageFragment + messageId.ToString(CultureInfo.InvariantCulture).PadLeft(3, '0');
                     var abpTasks = batch.Select(devices => SendUpstreamMessage(devices.Abp, payload, messageId));
```

---
PS You can copy [the above diff](https://github.com/Azure/iotedge-lorawan-starterkit/files/7949870/diff.txt) to the clipboard and apply it from PowerShell using `Get-Clipboard | git apply` (alternatively `type diff.txt | git apply` if you end up [downloading](https://github.com/Azure/iotedge-lorawan-starterkit/files/7949870/diff.txt) it).